src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp

```diff
@@ -1137,31 +1141,60 @@ struct MHAHelper {
+        size_t num_blocks_in_total = key_cache.size(0);
+        int32_t* rotated_block_indices_data = rotated_block_indices.ptr<int32_t>();
+        size_t num_heads = key_cache.size(1); // H;
+        size_t embedding_size = key_cache.size(3); // S;
+        size_t head_chunk_size = _block_size * embedding_size;
+
+        for (size_t i = 0; i < num_rotated_blocks; i++) {
+            size_t rotated_block_index = *(rotated_block_indices_data + i);
+            OPENVINO_ASSERT(rotated_block_index < num_blocks_in_total);
+            float* rotation_coefficient_block_data = rotation_coefficients.ptr<float>() + i * head_chunk_size;
+            KVCACHE_TYPE* cache_block_ptr = key_cache.ptr<KVCACHE_TYPE>(rotated_block_index);
+            rotate_kv_cache_block(cache_block_ptr, rotation_coefficient_block_data, num_heads, _block_size, embedding_size);
+
+        }
+    }
+
     // compute one token, loop along batch, head dimensions and kv_len, it's special for very long kv_len with small batch tokens.
     // It will assume NO mixture execution of first and second token.
     // all tensors such as query... have batch dimension which is DIFFERENT from above
     //  query: [B, H, L, S]
-    //  present_*: [block_number, H, 32, S]
+    //  key_cache: [block_number, H, _block_size, S]
+    //  value_cache: [block_number, H, _block_size, Sv]
     //  output_emb: [B, L, H * S]
+    //  rotated_block_indices: [num_rotated_blocks]
+    //  rotation_coefficients: [num_rotated_blocks, _block_size, S]
     // 3 loops along batch, head, kv cache length dimensions
     void exec_loop_bhl(const PlainTensor& query,
-                       const PlainTensor& present_key,
-                       const PlainTensor& present_value,
+                       PlainTensor& key_cache,
+                       PlainTensor& value_cache,
                        const PlainTensor& output_emb,
                        const PlainTensor& output_score,
                        size_t max_context_len,
                        const PlainTensor& past_lens,
                        const PlainTensor& subsequence_begins,
                        const PlainTensor& block_indices,
                        const PlainTensor& block_indices_begins,
-                       const PlainTensor& alibi_slopes) {
+                       const PlainTensor& alibi_slopes,
+                       const PlainTensor& rotation_coefficients,
+                       const PlainTensor& rotated_block_indices) {
         auto B = past_lens.size(0);
         auto q_len = query.size(2);
         auto kv_len_in_blocks = div_up(max_context_len, _block_size);
 
         // aligned to cache line (64bytes=16*sizeof(float)) to avoid false sharing
         _weight_bhl.resize<float>({B, _H, q_len, rnd_up(max_context_len, std::max(_block_size, size_t{16}))});
 
+        // TODO (vshampor): implement cache rotation at this spot
+        if (rotation_coefficients) {
```

Comment:
The logic should be also added for the more often used path `exec_loop_mixed`.

Reply:
Added, thanks!